### ISSUE TITLE
Forms Fields and Inputs strategy

### ISSUE THREAD

**Issue by landitus** [auth0/cosmos#177]
Looking it **from the documentation**, when looking at the TextInput, Select and Textarea components, these do not have labels. They are just the basic atoms or HTML element representation. The full field with the label and helper text can be accomplished within a Form, via the Form.TextInput component. 

Should the TextInput come with its label? Should we expose the TextInput to the docs when we do not provide a Label component? Would we want to use the Text Input with no labels? I think this is related to accessibility issues. An example of a input (outside of a Form) would be the Search in a header. Yet, it would be more accessible with a label (visually hidden).  

<img width="283" alt="screen shot 2018-01-12 at 08 47 00" src="https://user-images.githubusercontent.com/239215/34873844-74ad7f44-f775-11e7-83e8-92fc7072c2cf.png">

I think we touched briefly upon this, but I can't find a definitive strategy so we could clarify it here.
I've noticed some systems do it one way or the other. Our strategy would be similar to Bootstrap, but other systems provide the label with Text Input component, Textareas, etc. This could also be expanded to the Checkbox and Radio buttons components (which are rarely used).

**Comment by nkohari** [comment]
@landitus Is there an action to take here? I think this describes our current approach to form controls.

**Comment by jstrutz** [comment]
A couple notes on the things we've found, when working w/ forms for 2020:

### Surrounding a `<Form.TextInput>` element (and probably other types) with a non-presentational component breaks the form layout
```jsx
<Form layout="label-on-left">
   <>{/* Note the wrapping Fragment */}
   <Form.TextInput
     label="Email"
     name="email"
     type="text"
     placeholder="Your email address"
  />
  </>
</Form>
```
<img width="1057" alt="screen shot 2018-06-14 at 2 58 01 pm" src="https://user-images.githubusercontent.com/47222/41429380-5c8c3e50-6fe3-11e8-81b5-bc6bedaab7a2.png">

### Action Handlers need disabled state

So we can communicate that the form cannot be submitted (i.e. in the case of invalid input).  Example:

```jsx
<Form.Actions primaryAction={{
  disabled: true,
  label: 'Save Changes',
  handler: handleSubmit
}} />
```

**Comment by siddharthkp** [comment]
Found the culprit. We pass the layout prop to each child which breaks if you wrap them in a container.

```js
const Form = props => (
  <form>
    {React.Children.map(props.children, child =>
      React.cloneElement(child, { layout: props.layout })
    )}
  </form>
)
```

2 points here:

1. We can fix this by using the context API, it's the perfect use case (especially now that we are locking down at React 16.x)

2. @jstrutz What are you wrapping around individual form elements 🤔

**Comment by siddharthkp** [comment]
Here's the issue for the feature request: https://github.com/auth0/cosmos/issues/597

**Comment by siddharthkp** [comment]
Problem 1: The naming is not very easy to guess. Needs some work. I'll spin off a RFC and link it here

Problem 2: You should be able to attach actions and icons to TextInput, there is an issue here: https://github.com/auth0/cosmos/issues/1082

**Comment by siddharthkp** [comment]
Closing this PR as it's split into #1082 and #1190
